### PR TITLE
Throw exception when 10,000 rounds are hit in a battle

### LINKED ISSUE #7848: Infinite loop in battle calculator
### How can the problem be recreated?

Unsure to be honest. We believe that the properties for subs are corrupted in the save file and this later causes issues for attacking subs with defending planes but no defending destroyers. Probably vice versa too.

If you download the below file, Open the battle calculator for SZ112, swap sides, add in SZ111 defenders and some planes, then run the calculator, you will get the problem.

### Attach a Save Game

https://www.axisandallies.org/forums/assets/uploads/files/1600889975703-triplea_35795_1rus.tsvg

### If playing a prerelease, which version are you using?

Game Version:
2.2.20868

### If playing a prerelease, does this happen on the latest release?

- [] yes
- [] no

### Is there anything else we should know?
Hopefully, this is enough. Many players are noticing this problem. Also, subs lose their stealth properties at certain points; this is likely the same problem.

### PR DESCRIPTION
**Pull request by trevan**
Fixes #7848 

I think this is the last thing to fix for #7848.

## Testing

Ran world at war with Hard AI to turn 13 just to see if I'd hit this exception in normal game play.  The exception did not occur.

## Release Note

CHANGE|An exception will be thrown if a battle ever takes more than 10,000 rounds to help figure out how that happened.

### CODE REVIEW

**Comment by Cernelius** [comment]
I think 1,000 would be more than enough...

I assume this is a temporary patch, until that never actually happens.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java [hunk 1]
```diff
@@ -106,6 +106,8 @@
 
   private static final long serialVersionUID = 5879502298361231540L;
 
+  private static final long MAX_INFINITE_ROUNDS = 10000;
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:109]: What do you think of naming this `MAX_ROUNDS` instead?
"max infinite" does not quite grok well for me.

File: game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java [hunk 2]
```diff
@@ -1699,6 +1701,25 @@ public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
           public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
             if (!isOver) {
               round++;
+              if (round > MAX_INFINITE_ROUNDS) {
+                // the battle appears to be in an infinite loop
+                throw new IllegalStateException(
+                    "Round 10,000 reached in a battle. Something must be wrong."
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java:1707]: What do you think of adding to the message "Please report this to TripleA"?
A line break between this first line and the unit listing could make it an easier message to read as well.

**Review by DanVanAtta — COMMENTED** [review]
@trevan couple comments for you to consider.

**Comment by trevan** [comment]
@DanVanAtta , I incorporated your feedback.

**Comment by DanVanAtta** [comment]
@Cernelius I suspect you're right with 1k, at 10k we are quite confident though we do not have a false positive.

I don't expect this to be temporary, like *many* error conditions we have to handle when coding, this is one we never expect to hit but need to add handling if somehow we do (otherwise the game freezes).

**Comment by DanVanAtta** [comment]
@trevan would you mind temporarily & locally setting the max round to a low number (1 or 0) and verifying that the output format looks good?

**Comment by trevan** [comment]
Screen Shot 2020-10-28 at 10 22 54 PM

Here's what it looks like in the error dialog.  And I submitted it to github to see what it looks like here.  That issue is #7984.

**Comment by DanVanAtta** [comment]
Thank you for double checking @trevan  - looks good :+1:

**Comment by runspired** [comment]
If we're hitting this fairly consistently what info would best help y'all track it down?

**Comment by trevan** [comment]
@runspired , if you can give me a save right before it is happening, that would be extremely helpful.  If not, if you could click the "Report to TripleA" and then find the issue that was added to github and link to it from here.

**Comment by runspired** [comment]
@trevan I don't have a save but I'll start saving every turn so that I do for you on the next one, this seems to end 4 of 5 games for me at the moment.  I suspect it's because I make fairly copious use of edit-mode for the scenario I play to keep the sides balanced (the lotr map starts so unbalanced against the "evil" side that even with giving them sizable resource modifiers they lose pretty quick, so I usually bring Rhun back from the dead or some such).

**Pull request merged**